Use mp_obj_get_array where sequence may be a tuple or a list.

blazewicz opened the pull request:
This patch:
* reduces code size (-80 B on unix, -60 B on stmhal),
* fixes #2934.

blazewicz committed:
py: Use mp_obj_get_array where sequence may be a tuple or a list.

blazewicz committed:
tests/basics/string_join.py: Add test case where argument is not iterable.

blazewicz committed:
tests/basics/unpack1.py: Test if *a, = b copies b when b is a list.

dpgeorge commented:
Thanks, that's a really great bug find and fix.  But, please can you make a separate commit that fixes just the bug #2934 (since it's a bug fix and changes behaviour of the code), and then in another commit have the changes that make optimisations to reduce code size by using mp_obj_get_array.  And of course the tests should stay in different commits.  Thanks.

blazewicz committed:
py/runtime.c: Remove optimization of '*a,=b', it caused a bug.

*a, = b should always make a copy of b, instead, before this patch
if b was a list it would copy only a reference to it.

blazewicz commented:
I didn't do it in separate commits because the latter covers changes from the former, but you're right, its better to have it separate in commit history.

dpgeorge commented:
Merged, thank you very much!
